docs/architecture/clients/dependency-injection.md

```diff
@@ -0,0 +1,219 @@
+- `apps/<client-name>` for services used by a single client only
+
+If a service is used in an Angular context only, it can use the Angular `@Injectable` decorator to
+automatically configure its dependencies. If a service is used in mixed contexts (e.g.
+`libs/common`), it must not use Angular decorators and its dependencies must be manually configured.
+
+## Interfaces
+
+A service should implement an interface defined by an abstract class if:
+
+1. The service will be used across multiple client apps, or
+2. You want to define multiple interfaces for your service to limit what functionality is available
+   to different consumers
+
+For example, you may want to use interfaces to prevent outside modules from calling your internal
+methods:
+
+```ts
+// Injected in other modules
+abstract class MyService {
+  externalMethod: () => void;
+}
+
+// Injected in the MyService feature module
+abstract class InternalMyService extends MyService {
+  internalMethod: () => void;
+}
+
+class MyService implements InternalMyService {
+  externalMethod: () => void;
+  internalMethod: () => void;
+}
+```
+
+If neither of these apply to you, you probably don't need to define an abstract interface for your
+service.
+
+## Configuring DI
+
+### Angular modules
+
+Angular contexts generally use ngModules
+[dependency providers](https://angular.io/guide/dependency-injection-providers) to configure DI. For
+example:
+
+```ts
+@NgModule({
+  providers: [
+    // If the service does not have @Injectable decorators
+    {
+      provide: MyService, // abstract interface
+      useClass: DefaultMyService, // concrete implementation
+      deps: [DepA, DepB], // abstract interfaces for each dependency required by the DefaultMyService constructor
+    },
+    // If the service does have @Injectable decorators (Angular-only)
+    {
+      provide: SomeAngularService,
+      useClass: DefaultSomeAngularService,
+      // no deps array required
+    },
+  ],
+})
+export class MyServicesModule {}
+```
+
+By default, this is not type safe - Angular does not verify that the implementation actually
+implements the abstract interface, or that the `deps` array matches the constructor parameters.
+
+We provide a helper function called `safeProvider`, which acts as a wrapper around each provider
+entry and provides compile-time type checking of your configuration. For example (continuing with
+the example above):
+
+```ts
+@NgModule({
+  providers: [
+    safeProvider({
+      provide: MyService,
+      useClass: DefaultMyService,
+      deps: [DepA, DepB],
+    }),
+    safeProvider({
+      provide: SomeAngularService,
+      useClass: DefaultSomeAngularService,
+      useAngularDecorators: true, // this tells safeProvider to not require a deps array
+    }),
+  ],
+})
+export class MyServicesModule {}
+```
+
+:::note
+
+The `useAngularDecorators` option is not type safe - it is your responsibility to ensure that the
+class actually uses the `@Injectable` decorator.
+
+:::
```

Comment:
💭 thought: we should keep an eye on https://github.com/microsoft/TypeScript/issues/4881 which might be able to fix this

Reply:
Nice spot!

Reply:
Keeping this open for future reference! :)